core/src/main/java/org/ehcache/events/CacheEventDispatcherImpl.java

```diff
@@ -77,12 +89,30 @@ public CacheEventDispatcherImpl(ExecutorService orderedDelivery, ExecutorService
   @Override
   public void registerCacheEventListener(CacheEventListener<? super K, ? super V> listener,
                                   EventOrdering ordering, EventFiring firing, EnumSet<EventType> forEventTypes) {
-    boolean doRegister = forEventTypes.contains(EventType.EVICTED) || forEventTypes.contains(EventType.EXPIRED);
-    if (!registeredListeners.add(new EventListenerWrapper(listener, firing, ordering, forEventTypes))) {
-      throw new IllegalStateException("Cache Event Listener already registered: " + listener);
+    EventListenerWrapper wrapper = new EventListenerWrapper(listener, firing, ordering, forEventTypes);
+
+    registerCacheEventListener(wrapper);
+  }
+
+  private void registerCacheEventListener(EventListenerWrapper wrapper) {
+    if(aSyncListenersSet.contains(wrapper) || syncListenersSet.contains(wrapper)) {
+      throw new IllegalStateException("Cache Event Listener already registered: " + wrapper.listener);
     }
-    if (doRegister) {
-      store.enableStoreEventNotifications(storeListener);
+
+    if(wrapper.ordering == EventOrdering.ORDERED) {
```

Comment:
However, having an if/else further down for a two value enum - using switch there